Fix allocation cleaning deletion

sebageek opened the pull request:
The cleaning of unused allocations needs to be moved into the transaction to actually work.

We also improve the logging a bit by not logging each single allocation with two log lines and adding some debug logging to see how long each phase takes.

---
I looked into the sync code to improve the situation a bit and this was the most low hanging fruit I could find. Either reworking the logic or removing the sync loop altogether will still be in our future, though.

sven-rosenzweig reviewed (dismissed):
Looks good to me.

sebageek committed:
Switch to stestr for unit tests

sebageek committed:
Fix and improve allocation cleaning sync

This is a rewrite of the sync allocations algorithm, which should be
about 80% faster (due to different way to query the DB), hopefully
better to understand and should actually clean allocations up (as before
the deletion happened outside of a transaction and was never actually
committed).

I've started out to understand the original sync algorithm and why it is
so slow and arrived with a rewrite of it in the way. The rewrite ended
up being 5% faster (which is basically nothing in this context), but
hopefully it is better to understand. While profiling my code it came to
light that about 80% of the algorithm was spent creating all the
AllocationsModel instances. Querying only the fields that we need and
by that not creating the model instances makes it much faster. As we
don't need the instances anyway with the rewrite (we're doing most of
the aggregation via set operations) changing this works quite well.

The code now actually also has tests that it does what it should do.

In the long run we want to remove the custom allocation management from
the driver, which also should be possible as segment allocation in
Neutron is less (not?) prone to race conditions anymore since Yoga.
Transitioning without interrupting the ability to create new segments
and without introducing segmentation inconsistencies is not trivial
though, which is why we go with the sync rewrite before developing the
whole migration.

swagner-de commented on networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py:
```diff
@@ -348,70 +349,70 @@ def _segmentation_ids(host_config):
+                    if db_alloc.host not in config_hg_names:
+                        if db_alloc.host not in hosts_to_delete and db_alloc.network_id is None:
+                            # at least one allocation with this host can be deleted
+                            hosts_to_delete.append(db_alloc.host)
```
Is there a reason hosts_to_delete is not a set and we save a few checks?

sebageek replied:
That is a pretty good question. This could just be a `set()`.

swagner-de commented on networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py:
```diff
@@ -348,70 +349,70 @@ def _segmentation_ids(host_config):
+                    # delete extra allocs
+                    extra_ids = no_net_allocations[hg_name] - hg_vlans
+                    if extra_ids:
+                        LOG.debug("Deleting %s unused allocations for hostgroup %s", len(extra_ids), hg_name)
```
These are unused and according to the config also not in the allowed physnet range? If so, let the log line reflect it.

swagner-de commented on networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py:
```diff
@@ -348,70 +349,70 @@ def _segmentation_ids(host_config):
+
+            # remove allocs on orphaned hostgroups that don't have a network_id assigned
+            if hosts_to_delete:
+                LOG.debug("Deleting old hosts that have empty allocations: %s", ", ".join(hosts_to_delete))
```
Old hosts that are non-existing in the config and have no allocations anymore?

sebageek committed:
Only execute allocation sync once per pod/system

We now make sure that the allocations (and hostgroup modes) are only
synced once per system - in our main use-case this is once per
neutron-server pod in Kubernetes. This is done by having the first
process to start the sync create a file after it is done. If that file
is found, subsequent neutron-server processes will not initiate the sync
on the same system. To make sure we don't end up in another race
condition with this we use a external_lock() from oslo.concurrency,
which guards the critical code section.

sven-rosenzweig commented on networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py:
```diff
@@ -42,7 +44,29 @@
+                self._sync_db()
+                with open(CONF.ml2_aci.sync_allocations_done_file_path, "w") as f:
+                    f.write(str(os.getpid()))
+                LOG.info("Successfully synced db")
```
Does it make to track the sync time?

sebageek replied:
Hmm you're thinking about tracking the time for the whole call? We could do that. Currently you could look at the two timestamps and subtract the time, but a direct output for it might be nice.

sven-rosenzweig commented on networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py:
```diff
@@ -42,7 +44,29 @@
+                synced_by = open(CONF.ml2_aci.sync_allocations_done_file_path, "r").read()
+                LOG.info("Not syncing in this thread, already synced by PID %s", synced_by)
+            except FileNotFoundError:
+                self._sync_db()
```
Comment like 'First thread gathering the lock, starts the sync' would be helpful

sven-rosenzweig commented on networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py:
```diff
@@ -348,70 +349,72 @@ def _segmentation_ids(host_config):
+            # sort allocations into data structures
+            LOG.debug("Processing allocations")
+            allocations = {}  # vlan ids grouped by hostgroup/physnet name
+            no_net_allocations = {}  # same as allocations for allocs that don't have a network assigned
```
` # same as allocations for allocs that don't have a network assigned` --> ` # vland ids grouped by hostgroup/physnet that don't have a network assigned`

sven-rosenzweig commented on networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py:
```diff
@@ -348,70 +349,72 @@ def _segmentation_ids(host_config):
+                hg_vlans = self._segmentation_ids(hg_config)
+                if hg_name in allocations:
+                    # delete extra allocs
+                    extra_ids = no_net_allocations[hg_name] - hg_vlans
```
does it make sense to name `extra_ids` to `out_of_ragnge_..`

sven-rosenzweig commented:
PR description is not matching what is happening in the commits.

sven-rosenzweig commented on networking_aci/tests/unit/plugins/ml2/drivers/mech_aci/test_allocations_manager.py:
```diff
@@ -0,0 +1,129 @@
+            for host, seg_id, net_id in start_data:
+                sess.add(AllocationsModel(host=host, level=1, segment_type="vlan", segmentation_id=seg_id,
+                                          network_id=net_id))
+            sess.add(AllocationsModel(host="sentinel_1", level=2, segment_type="vlan", segmentation_id=1))
```
comment why those two entries stay would be helpful. `not touched by sync algorithm due to binding level and segment tyep`

sven-rosenzweig reviewed (dismissed):
i like the code improvements.
